File: Iter2/src/Department.java
//Read Json & JSONParser
//Add Lecturers
//Add Teaching Assistants
//Add Courses

//Generate random capacity for course

//Add Course Prerequesties

//Add given course to correct course list (NTE-TE-UE-FTE)

//Keep record of how many students failed to register courses

//Reset capacity of all courses after each semester

//Create JSON object for department

import java.util.ArrayList;
import java.util.HashMap;

public class Department {

  private String name;
  private String id;
  private String faculty;
  private String currentSemester;

  private int departmentNumber;
  private int numberOfLectureDays;
  private int numberOfLectureHours;
  private int maximumCredits;
  private int courseTakeLimit;
  private int labCapacityLimit;
  private int electiveCapacityLimit;

  private Curriculum curriculum;

  private HashMap<String, DepartmentCourse> courses;
  private HashMap<String, DepartmentCourse> mandatoryCourses;
  private HashMap<String, DepartmentCourse> technicalElectiveCourses;
  private HashMap<String, DepartmentCourse> nonTechnicalElectiveCourses;
  private HashMap<String, DepartmentCourse> facultyElectiveCourses;
  private HashMap<String, DepartmentCourse> universityElectiveCourses;

  private ArrayList<Lecturer> lecturers;
  private ArrayList<TeachingAssistant> teachingAssistants;
  private ArrayList<Student> students;

  private final HashMap<String, HashMap<String, Integer>> courseRegistrationMessages;
  private final String[] registrationMessageFormats = {
      "%s students couldn't register %s because of prerequisites did not ok.",
      "%s students couldn't register %s because there is a collision with another course.",
      "%s students couldn't register %s because credit limit exceeded.",
      "%s students couldn't register %s because of quota problem.",
      "%s students couldn't register %s because of quota problem.",
      "%s students couldn't register %s because it has collision with another course.",
      "%s students couldn't register %s because it doesn't include for current semester.",
      "%s students couldn't register %s because they did not complete minimum credits.",
  };

  public Department(String name, String id, String faculty) {
    this.name = name;
    this.id = id;
    this.faculty = faculty;
    lecturers = new ArrayList<>();
    teachingAssistants = new ArrayList<>();
    students = new ArrayList<>();
    courses = new HashMap<>();
    mandatoryCourses = new HashMap<>();
    technicalElectiveCourses = new HashMap<>();
    nonTechnicalElectiveCourses = new HashMap<>();
    facultyElectiveCourses = new HashMap<>();
    universityElectiveCourses = new HashMap<>();
    courseRegistrationMessages = new HashMap<>();
    curriculum = new Curriculum();
  }

  public DepartmentCourse createCourse(String courseId, String courseName, String courseType, int credits,
      int lectureHours, int semester, int capacity) {

    DepartmentCourse courseObject = new DepartmentCourse();

    switch (courseType) {
      case "TE":
      case "UE":
      case "NTE":
      case "FTE":
        courseObject = new ElectiveCourse(courseId, courseName, credits, semester, lectureHours, capacity, courseType);
        break;
      case "M":
        courseObject = new MandatoryCourse(courseId, courseName, credits, semester, lectureHours);
        break;
    }
    
    return courseObject;
  }

  public void addCourse(DepartmentCourse course) {
    if (course.getSemester() == 0 && course instanceof ElectiveCourse) {
      switch (((ElectiveCourse) course).getType()) {
        case "TE":
          technicalElectiveCourses.put(course.getCourseId(), course);
          break;
        case "UE":
          universityElectiveCourses.put(course.getCourseId(), course);
          break;
        case "NTE":
          nonTechnicalElectiveCourses.put(course.getCourseId(), course);
          break;
        case "FTE":
          facultyElectiveCourses.put(course.getCourseId(), course);
          break;
      }
      this.courses.put(course.getCourseId(), course);
    } else if (course instanceof MandatoryCourse) {
      mandatoryCourses.put(course.getCourseId(), course);
      this.courses.put(course.getCourseId(), course);
      this.curriculum.addCourse(course.getSemester(), course);
    } else if (course instanceof ElectiveCourse && course.getSemester() != 0) {
      this.curriculum.addCourse(course.getSemester(), course);
    }
  }

  public void resetCapacities() {
    for (DepartmentCourse course : this.courses.values()) {
      if (course instanceof MandatoryCourse && course.getLabSections().size() > 0) {
        for (LabCourse labSection : course.getLabSections()) {
          int randomSectionCapacity = (int) (Math.random() * labCapacityLimit) + 1;
          labSection.setCapacity(randomSectionCapacity);
        }
      } else if (course instanceof ElectiveCourse) {
        int randomCapacity = (int) (Math.random() * electiveCapacityLimit) + 1;
        ((ElectiveCourse) course).setCapacity(randomCapacity);
      }
    }
  }

  public DepartmentCourse getRandomCourse(Object[] courses) {
    int randomIndex = (int) (Math.random() * courses.length);
    return (DepartmentCourse) courses[randomIndex];
  }

  public DepartmentCourse getCourse(String courseId) {
    return this.courses.get(courseId);
  }

  public String getName() {
    return name;
  }

  public void setName(String name) {
    this.name = name;
  }

  public String getId() {
    return id;
  }

  public void setId(String id) {
    this.id = id;
  }

  public int getCourseTakeLimit() {
    return courseTakeLimit;
  }

  public void setCourseTakeLimit(int courseTakeLimit) {
    this.courseTakeLimit = courseTakeLimit;
  }

  public String getCurrentSemester() {
    return currentSemester;
  }

  public void setCurrentSemester(String currentSemester) {
    this.currentSemester = currentSemester;
  }

  public Curriculum getCurriculum() {
    return curriculum;
  }

  public ArrayList<Lecturer> getLecturers() {
    return lecturers;
  }

  public Lecturer getLecturer(int id) {
    return lecturers.get(id);
  }

  public void addLecturer(String name, Department department) {
    lecturers.add(new Lecturer(name, department));
  }

  public void addTeachingAssistant(String name, Department department) {
    teachingAssistants.add(new TeachingAssistant(name, department));
  }

  public void addStudent(Student student) {
    students.add(student);
  }

  public String getFaculty() {
    return faculty;
  }

  public void setFaculty(String faculty) {
    this.faculty = faculty;
  }

  public void setLabCapacityLimit(int labCapacityLimit) {
    this.labCapacityLimit = labCapacityLimit;
  }

  public void setElectiveCapacityLimit(int electiveCapacityLimit) {
    this.electiveCapacityLimit = electiveCapacityLimit;
  }

  public int getNumberOfLectureDays() {
    return numberOfLectureDays;
  }

  public void setNumberOfLectureDays(int numberOfLectureDays) {
    if (numberOfLectureDays > 7 || numberOfLectureDays < 1) {
      numberOfLectureDays = 5;
    }
    this.numberOfLectureDays = numberOfLectureDays;
  }

  public int getNumberOfLectureHours() {
    return numberOfLectureHours;
  }

  public void setNumberOfLectureHours(int numberOfLectureHours) {
    if (numberOfLectureHours > 24 || numberOfLectureHours < 1) {
      numberOfLectureHours = 15;
    }
    this.numberOfLectureHours = numberOfLectureHours;
  }

  public int getDepartmentNumber() {
    return departmentNumber;
  }

  public void setDepartmentNumber(int departmentNumber) {
    this.departmentNumber = departmentNumber;
  }

  public HashMap<String, DepartmentCourse> getCourses() {
    return courses;
  }

  public HashMap<String, DepartmentCourse> getMandatoryCourses() {
    return mandatoryCourses;
  }

  public HashMap<String, DepartmentCourse> getTechnicalElectiveCourses() {
    return technicalElectiveCourses;
  }

  public HashMap<String, DepartmentCourse> getNonTechnicalElectiveCourses() {
    return nonTechnicalElectiveCourses;
  }

  public HashMap<String, DepartmentCourse> getFacultyElectiveCourses() {
    return facultyElectiveCourses;
  }

  public HashMap<String, DepartmentCourse> getUniversityElectiveCourses() {
    return universityElectiveCourses;
  }

  public ArrayList<Student> getStudents() {
    return students;
  }

  public ArrayList<TeachingAssistant> getTeachingAssistants() {
    return teachingAssistants;
  }

  public TeachingAssistant getTeachingAssistant(int index) {
    return teachingAssistants.get(index);
  }

  public int getLabCapacityLimit() {
    return labCapacityLimit;
  }

  public int getElectiveCapacityLimit() {
    return electiveCapacityLimit;
  }

  public int getMaximumCredits() {
    return maximumCredits;
  }

  public void setMaximumCredits(int maximumCredits) {
    this.maximumCredits = maximumCredits;
  }

  public void setCurriculum(Curriculum curriculum) {
    this.curriculum = curriculum;
  }

  public void addCourseRegistrationMessage(String messageFormat, Course course) {
    if (!courseRegistrationMessages.containsKey(messageFormat)) {
      courseRegistrationMessages.put(messageFormat, new HashMap<>());
    } else {
      courseRegistrationMessages.get(messageFormat).merge(course.getCourseId(), 1, Integer::sum);
    }
  }

  public void addCourseRegistrationMessage(int messageId, Course course) {
    String messaageFormat = registrationMessageFormats[messageId];
    if (!courseRegistrationMessages.containsKey(messaageFormat)) {
      courseRegistrationMessages.put(messaageFormat, new HashMap<>());
    } else {
      courseRegistrationMessages.get(messaageFormat).merge(course.getCourseId(), 1, Integer::sum);
    }
  }

  public HashMap<String, HashMap<String, Integer>> getCourseRegistrationMessages() {
    return courseRegistrationMessages;
  }

}

File: Iter3/src/department.py
import random

from curriculum import Curriculum

import person
import course
from department_course import DepartmentCourse
from elective_course import ElectiveCourse
from mandatory_course import MandatoryCourse


class Department:
    __name = str
    __id = str
    __faculty = str
    __current_semester = str
    __department_number = int
    __number_of_lecture_days = int
    __number_of_lecture_hours = int
    __maximum_credits = int
    __curriculum = Curriculum
    __courses = {}
    __mandatory_courses = {}
    __technical_elective_courses = {}
    __non_technical_elective_courses = {}
    __faculty_elective_courses = {}
    __university_elective_courses = {}
    __lecturers = []
    __teaching_assistants = []
    __students = []
    __course_take_limit = int
    __lab_capacity_limit = int
    __elective_capacity_limit = int
    __course_registration_messages = {}
    __registration_message_formats = [
        "%s students couldn't register {} because of prerequisites did not met.",
        "%s students couldn't register {} because it has collision with another course.",
        "%s students couldn't register {} because they are reached credit limit.",
        "%s students couldn't register {} because of quota problem.",
        "%s students couldn't register {} because of quota problem.",
        "%s students couldn't register {} because it has collision with another course.",
        "%s students couldn't register {} because the course belongs to another semester.",
        "%s students couldn't register {} because they did not complete minimum credits.",
    ]

    def __init__(self, name, id, faculty, current_semester, department_number, number_of_lecture_days,
                 number_of_lecture_hours):
        self.__name = name
        self.__id = id
        self.__faculty = faculty
        self.__current_semester = current_semester
        self.__department_number = department_number
        self.__number_of_lecture_days = number_of_lecture_days
        self.__number_of_lecture_hours = number_of_lecture_hours

    def create_course(self, course: dict):
        semester = 0
        if course['semester'] is not None:
            semester = int(course['semester'])

        random_capacity = int(
            (random.uniform(0, self.__elective_capacity_limit) + 1))

        if course['type'] == "M":
            course_object = MandatoryCourse(course['courseId'], course['courseName'], credits, semester,
                                            course['lectureHours'])
        else:
            course_object = ElectiveCourse(course['courseId'], course['courseName'], credits, semester,
                                           course['lectureHours'], random_capacity, course['type'])

        if course.get("minCompletedCredits") is not None:
            course_object.set_min_completed_credits(
                int(course['minCompletedCredits']))

        return course_object

    def add_course(self, course):
        if course.get_semester() == 0 and isinstance(course, ElectiveCourse):
            course_type = course.get_type()
            self.__courses[course.get_course_id()] = course
        elif isinstance(course, MandatoryCourse):
            self.__mandatory_courses[course.get_course_id()] = course
            self.__courses[course.get_course_id()] = course
            self.__curriculum.add_course(int(course.get_semester()), course)
        elif isinstance(course, ElectiveCourse) and course.get_semester() != 0:
            self.__curriculum.add_course(course.get_semester(), course,)

    def add_course_registration_message(self, message_id, course):
        message_format = self.__registration_message_formats[message_id]
        if message_format not in self.__course_registration_messages:
            self.__course_registration_messages.extend(message_format)
        else:
            self.__course_registration_messages.get(
                message_format).merge(course.get_course_id())

    def reset_capacities(self):
        for course in self.__courses:
            if isintance(course, MandatoryCourse) and len(course.get_lab_sections()) > 0:
                for lab_section in course.get_lab_sections():
                    random_section_capacity = int(
                        random.uniform(1, self.__lab_capacity_limit) + 1)
            elif isintance(course, ElectiveCourse):
                random_capacity = int(random.uniform(
                    1, self.__elective_capacity_limit) + 1)
                ElectiveCourse(course).set_capacity(random_capacity)

    def get_name(self):
        return self.__name

    def set_name(self, name):
        self.__name = name

    def get_id(self):
        return self.__id

    def set_id(self, id):
        self.__id = id

    def get_course_take_limit(self):
        return self.__course_take_limit

    def get_current_semester(self):
        return self.__current_semester

    def get_curriculum(self):
        return self.__curriculum

    def get_lecturer(self, id: int):
        return self.__lecturers[id]

    def add_lecturer(self, lecturer):
        self.__lecturers.append(lecturer)

    def add_teaching_assistant(self, assistant):
        self.__teaching_assistants.append(assistant)

    def add_student(self, student):
        self.__students.append(student)

    def get_faculty(self):
        return self.__faculty

    def get_technical_elective_courses(self):
        return self.__technical_elective_courses

    def get_non_technical_elective_courses(self):
        return self.__non_technical_elective_courses

    def get_faculty_elective_courses(self):
        return self.__faculty_elective_courses

    def get_university_elective_courses(self):
        return self.__university_elective_courses

    def get_number_of_lecture_days(self):
        return self.__number_of_lecture_days

    def get_number_of_lecture_hours(self):
        return self.__number_of_lecture_hours

    def get_department_number(self):
        return self.__department_number

    def get_courses(self):
        return self.__courses

    def get_students(self):
        return self.__students

    def get_course_registration_messages(self):
        return self.__course_registration_messages

    def get_maximum_credits(self):
        return self.__maximum_credits

    def get_elective_capacity_limit(self):
        return self.__elective_capacity_limit

    def set_course_take_limit(self, course_take_limit: int):
        self.__course_take_limit = course_take_limit

    def set_lab_capacity_limit(self, lab_capacity_limit: int):
        self.__lab_capacity_limit = lab_capacity_limit

    def set_elective_capacity_limit(self, elective_capacity_limit: int):
        self.__elective_capacity_limit = elective_capacity_limit

    def set_maximum_credits(self, maximum_credits: int):
        self.__maximum_credits = maximum_credits
